### ISSUE TITLE
Incompatible shapes for one-hot encoded multiclass targets (SVGP)

### ISSUE THREAD

**Issue by AtheMathmo** [GPflow/GPflow#438]
I'm writing this issue mostly to propose clearer documentation around this (and to confirm my understanding).

The SVGP class description has:

```
- X is a data matrix, size N x D
- Y is a data matrix, size N x R
...
- num_latent is the number of latent process to use, default to Y.shape[1]
...
```

However, if I feed in a one-hot encoded data matrix Y, of size `(N, 10)` for example, I receive the following runtime error:

```
InvalidArgumentError (see above for traceback): Incompatible shapes: [500,10] vs. [50,10]
         [[Node: model.build_likelihood/model.likelihood.variational_expectations/mul = Mul[T=DT_DOUBLE, _device="/job:localhost/replica:0/task:0/cpu:0"](model.build_likelihood/model.likelihood.variational_expectations/Cast, model.build_likelihood/add)]]
```

(Full stack trace at end)

Looking at the tutorial, I believe that the model actually expects an `Nx1` matrix with each entry corresponding to the class label. Internally it seems that the model does a [one-hot encoding later](https://github.com/GPflow/GPflow/blob/ad168b3bb0ab5278f6d5f1738b14d4c10eccd10f/GPflow/likelihoods.py#L376-L398).

Therefore it is necessary to specify the number of latent processes also?

---

Full stack trace:

```
File "DIR\model.py", line 224, in optimize
    return self._optimize_tf(method, callback, maxiter, **kw)
  File "DIR\model.py", line 230, in _optimize_tf
    opt_step = self._compile(optimizer=method)
  File "DIR\model.py", line 133, in _compile
    f = self.build_likelihood() + self.build_prior()
  File "DIR\scoping.py", line 41, in runnable
    return f(*args, **kwargs)
  File "DIR\svgp.py", line 106, in build_likelihood
    var_exp = self.likelihood.variational_expectations(fmean, fvar, self.Y)
  File "DIR\scoping.py", line 41, in runnable
    return f(*args, **kwargs)
  File "DIR\likelihoods.py", line 429, in variational_expectations
    p = self.invlink.prob_is_largest(Y, Fmu, Fvar, gh_x, gh_w)
  File "DIR\likelihoods.py", line 379, in prob_is_largest
    mu_selected = tf.reduce_sum(oh_on * mu, 1)
  File "TF_DIR\tensorflow\lib\site-packages\tensorflow\python\ops\math_ops.py", line 814, in binary_op_wrapper
    return func(x, y, name=name)
  File "TF_DIR\tensorflow\lib\site-packages\tensorflow\python\ops\math_ops.py", line 987, in _mul_dispatch
    return gen_math_ops.mul(x, y, name=name)
  File "TF_DIR\tensorflow\lib\site-packages\tensorflow\python\ops\gen_math_ops.py", line 1613, in mul
    result = _op_def_lib.apply_op("Mul", x=x, y=y, name=name)
  File "TF_DIR\tensorflow\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 759, in apply_op
    op_def=op_def)
  File "TF_DIR\tensorflow\lib\site-packages\tensorflow\python\framework\ops.py", line 2240, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "TF_DIR\tensorflow\lib\site-packages\tensorflow\python\framework\ops.py", line 1128, in __init__
    self._traceback = _extract_stack()
```

**Comment by hughsalimbeni** [comment]
I think there's some inconsistency here which should be cleared up: when using the bernoulli likelihood you need to pass (N, 1), but for the multiclass it needs to be shape (N, ) as the multiclass likelihood does a one-hot itself. Also, the input needs to be ints, so predict_density won't work unless you modify it like this: 

```python
class MultiClassSVPG(SVGP):
    @AutoFlow((tf.float64, [None, None]), (tf.int32, [None,]))
    def predict_density(self, Xnew, Ynew):
        pred_f_mean, pred_f_var = self.build_predict(Xnew)
        return self.likelihood.predict_density(pred_f_mean, pred_f_var, Ynew)
```

which is taken from here https://github.com/ICL-SML/Doubly-Stochastic-DGP/blob/master/demos/demo_mnist.ipynb

**Comment by jameshensman** [comment]
Woah, multiclass needs to be (N,) ? That's definitely a bug. I think it
should be (N, 1), in integer encoding,

The modification to predict_density is unfortunate, it's because we use the
same autoflow for all likelihoods which just assumes floats. Perhaps we
could just cast to int inside the likelihood, Hugh?

**Comment by imsrgadich** [comment]
Hi all, can you guys please help me with [this](https://github.com/ICL-SML/Doubly-Stochastic-DGP/issues/16). Thank you.
